app_dart/lib/src/service/luci_build_service_v2.dart

```diff
@@ -0,0 +1,1008 @@
+import '../model/appengine/commit.dart';
+import '../model/appengine/task.dart';
+import '../model/firestore/commit.dart' as firestore_commit;
+import '../model/firestore/task.dart' as firestore;
+import '../model/ci_yaml/target.dart';
+import '../model/github/checks.dart' as cocoon_checks;
+import '../model/luci/buildbucket.dart'; // targets use their own RequestedDimension which is defined here for some reason.
+import '../model/luci/user_data.dart';
+import '../service/datastore.dart';
+import '../service/logging.dart';
+import 'build_bucket_v2_client.dart';
+import 'exceptions.dart';
+import 'github_service.dart';
+
+/// Class to interact with LUCI buildbucket to get, trigger
+/// and cancel builds for github repos. It uses [config.luciTryBuilders] to
+/// get the list of available builders.
+class LuciBuildServiceV2 {
+  LuciBuildServiceV2({
+    required this.config,
+    required this.cache,
+    required this.buildBucketV2Client,
+    GithubChecksUtil? githubChecksUtil,
+    GerritService? gerritService,
+    this.pubsub = const PubSub(),
+  })  : githubChecksUtil = githubChecksUtil ?? const GithubChecksUtil(),
+        gerritService = gerritService ?? GerritService(config: config);
+
+  BuildBucketV2Client buildBucketV2Client;
+  final CacheService cache;
+  Config config;
+  GithubChecksUtil githubChecksUtil;
+  GerritService gerritService;
+
+  final PubSub pubsub;
+
+  static const Set<bbv2.Status> failStatusSet = <bbv2.Status>{
+    bbv2.Status.CANCELED,
+    bbv2.Status.FAILURE,
+    bbv2.Status.INFRA_FAILURE,
+  };
+
+  static const int kBackfillPriority = 35;
+  static const int kDefaultPriority = 30;
+  static const int kRerunPriority = 29;
+
+  /// Github labels have a max length of 100, so conserve chars here.
+  /// This is currently used by packages repo only.
+  /// See: https://github.com/flutter/flutter/issues/130076
+  static const String githubBuildLabelPrefix = 'override:';
+  static const String propertiesGithubBuildLabelName = 'overrides';
+
+  /// Name of the subcache to store luci build related values in redis.
+  static const String subCacheName = 'luci';
+
+  // the Request objects here are the BatchRequest object in bbv2.
+  /// Shards [rows] into several sublists of size [maxEntityGroups].
+  Future<List<List<bbv2.BatchRequest_Request>>> shard({
+    required List<bbv2.BatchRequest_Request> requests,
+    required int maxShardSize,
+  }) async {
+    final List<List<bbv2.BatchRequest_Request>> shards = [];
+    for (int i = 0; i < requests.length; i += maxShardSize) {
+      shards.add(requests.sublist(i, i + min<int>(requests.length - i, maxShardSize)));
+    }
+    return shards;
+  }
+
+  Future<Iterable<bbv2.Build>> getTryBuilds({
+    required github.RepositorySlug slug,
+    required String sha,
+    String? builderName,
+  }) async {
+    final List<bbv2.StringPair> tags = [
+      bbv2.StringPair(
+        key: 'buildset',
+        value: 'sha/git/$sha',
+      ),
```

Comment:
How does v2 handle tags where the key has multiple values, like the old one `buildset: <String>['sha/git/sha', 'buildset2']`?  Or we are allowed to have multiples StringPairs with same `key`?

Reply:
You got it right. There will be multiple StringPair objects with the same key.